pkgs/by-name/in/install-grub-ng/src/builder/install.rs

```diff
@@ -0,0 +1,456 @@
+use std::{
+    collections::HashSet,
+    fs,
+    io::{BufRead, BufReader, BufWriter, Write},
+    os::unix::fs::symlink,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+use eyre::{Context, Result, bail};
+
+use super::Builder;
+use crate::config::Config;
+
+impl Builder<'_> {
+    pub fn install(&mut self) -> Result<&mut Self> {
+        let efi_target = EfiTarget::deduce(&self.config)?;
+        let conf = self.config.boot_path.join("grub/grub.cfg");
+        let temp = self.config.boot_path.join("grub/grub.cfg.tmp");
+
+        if self.dry_run {
+            println!("{}", self.inner);
+            return Ok(self);
+        }
+
+        fs::write(&temp, &self.inner)?;
+
+        self.append_prepare_config()?;
+        self.run_os_prober(&efi_target, &temp)?;
+
+        // Atomically switch to the new config
+        fs::rename(&temp, &conf)
+            .with_context(|| format!("Cannot rename {} to {}", temp.display(), conf.display()))?;
+
+        self.remove_old_kernels()?;
+
+        let mut grub_state = GrubState::load(&self.config);
+
+        if grub_state.update(&self.config, &efi_target) {
+            if std::env::var("NIXOS_INSTALL_GRUB").as_deref() == Ok("1") {
+                eprintln!("NIXOS_INSTALL_GRUB env var deprecated, use NIXOS_INSTALL_BOOTLOADER");
+                std::env::set_var("NIXOS_INSTALL_BOOTLOADER", "1");
+            }
+
+            self.install_bios(&efi_target)?;
+            self.install_efi(&efi_target)?;
+
+            grub_state.save()?;
+        };
+
+        Ok(self)
+    }
+
+    fn append_prepare_config(&self) -> Result<()> {
+        let extra_prepare_config = self
+            .config
+            .extra_prepare_config
+            .replace("@bootPath@", &self.config.boot_path.to_string_lossy());
```

Comment:
The "lossy" `Path` funcs shouldn't be used, especially when we're already in a function returning a `Result`.
And we don't need to allocate a `String` to then use it as a `&str`:
```suggestion
        let boot_path = self
                .config
                .boot_path
                .to_str()
                .ok_or_else(|| eyre!("boot path is not valid UTF-8: {}", ))?;

        let extra_prepare_config = self
            .config
            .extra_prepare_config
            .replace("@bootPath@", boot_path);
```